### ISSUE TITLE
Thoughts on blending

### ISSUE THREAD

**Issue by raphlinus** [googlefonts/colr-gradients-spec#360]
As piet-gpu is implementing the blending nodes (`PaintComposite`) we've run into a few issues. None of these are showstoppers, but I'm documenting them here for consideration, possibly to improve quality and performance for a future revision.

First, each `PaintComposite` requires in the general case two temporary image pixels to be allocated and two blends. The first gives a clean canvas in which to paint backdrop, and the second similarly for src. The latter two are blended with the specified blend or composite, and the result is composited with Porter-Duff "over" to what I'll be calling the "underlayer", ie the paint nodes that have preceded this. This structure appears to be inherited from SVG, which has similar semantics for [feBlend](https://www.w3.org/TR/SVG11/filters.html#feBlendElement) and [feComposite](https://www.w3.org/TR/SVG11/filters.html#feCompositeElement).

Strictly from an efficiency point of view, it might be worth considering an approach more like HTML5 canvas than SVG, which would facilitate similar composition effects with a single temporary image pixel and single compositing operation.  In a strawman proposal, I would suggest this node has a *single* child, and the compositing happens against the underlayer with the specified blend operation. Further, the newly allocated image pixel might be clear (isolated) or duplicated from the underlayer, depending on a flags.

Obviously it would still be possible to represent SVG semantics by having two such nodes. In many cases, though, I think it should be possible to generate a more optimized representation.

Second, the concept of blend-group opacity is missing. In the current format, I believe the normal way to represent this is a `PaintComposite` with a src-in composition mode and a constant color with some opacity as the dst. 

Finally, a potentially sensitive point to potentially revisit is the colorspace for such compositing. Tradition and existing tools specify sRGB, but there are documented advantages to linear intensity, see for example Figure 8 of [What Every Programmer Should Know About Gamma](https://blog.johnnovak.net/2016/09/21/what-every-coder-should-know-about-gamma/). When doing exact-area antialiasing (as is appropriate for font rendering), edges are smoother with linear intensity than doing antialiasing in sRGB, for similar reasons.

In piet-gpu, we'll work around these in various ways, possibly implementing optimizations to rewrite more expensive blending operations into more efficient ones, when it's possible to infer these automatically without impacting rendering accuracy. But it might be worth revisiting in a future milestone.

**Comment by behdad** [comment]
> First, each `PaintComposite` requires in the general case two temporary image pixels to be allocated and two blends. The first gives a clean canvas in which to paint backdrop, and the second similarly for src. The latter two are blended with the specified blend or composite, and the result is composited with Porter-Duff "over" to what I'll be calling the "underlayer", ie the paint nodes that have preceded this. This structure appears to be inherited from SVG, which has similar semantics for [feBlend](https://www.w3.org/TR/SVG11/filters.html#feBlendElement) and [feComposite](https://www.w3.org/TR/SVG11/filters.html#feCompositeElement).

The way I envisioned this, there's no "underlayer". The graph is the entirety of the drawing.

> Second, the concept of blend-group opacity is missing. In the current format, I believe the normal way to represent this is a `PaintComposite` with a src-in composition mode and a constant color with some opacity as the dst.

Indeed, @rsheeter wanted a paint-with-alpha operation out on the basis that it can be achieved already and is not hugely needed on byte-saving basis.

> Finally, a potentially sensitive point to potentially revisit is the colorspace for such compositing. Tradition and existing tools specify sRGB, but there are documented advantages to linear intensity, see for example Figure 8 of [What Every Programmer Should Know About Gamma](https://blog.johnnovak.net/2016/09/21/what-every-coder-should-know-about-gamma/). When doing exact-area antialiasing (as is appropriate for font rendering), edges are smoother with linear intensity than doing antialiasing in sRGB, for similar reasons.

I agree that correct composition should happen in linear space, but that rarely happens.

**Comment by behdad** [comment]
> and the result is composited with Porter-Duff "over" to what I'll be calling the "underlayer", ie the paint nodes that have preceded this.

That's yet another node. That's what I'm trying to say.

There's one type of node that is a a "list of nodes over eachover". That's what you are referring to as underlayer here. But that's not necessarily always the parent of your PaintComposite.

In general I think for the PaintComposite it's possible to reuse the dst buffer as the output buffer. What the parent node does is a different issue.

**Comment by behdad** [comment]
> There's one type of node that is a a "list of nodes over eachover". That's what you are referring to as underlayer here. But that's not necessarily always the parent of your PaintComposite.

`PaintColrLayers`. That node can keep one buffer and accumulate as it goes.

**Comment by rsheeter** [comment]
Thank you for taking the time to write this up.

> two temporary image pixels to be allocated

Here's the current PaintComposite and pseudocode (from https://github.com/googlefonts/colr-gradients-spec):

```
struct PaintComposite
{
  uint8                  format; // = 32
  Offset24<Paint>        sourcePaint;
  CompositeMode          compositeMode;   // If mode is unrecognized use COMPOSITE_CLEAR
  Offset24<Paint>        backdropPaint;
};

// pseudo-code current
          saveLayer()
          paint Paint for backdrop, call a)
          saveLayer() with setting composite mode on SkPaint
          paint Paint for src, call a)
          restore with saved composite mode
          restore the outer layer
```

As Raph noted 1:1, this matches the Skia [implementation](https://source.chromium.org/chromium/chromium/src/+/master:third_party/skia/src/ports/SkFontHost_FreeType_common.cpp;drc=0bbce33b7ccbe1534cd001e538fa51f4ec008d1d;l=968).

Am I correct in thinking the proposed new paint looks something like this:

```
struct PaintComposite2
{
  uint8                  format; // = TBD
  Offset24<Paint>        paint;
  CompositeMode          compositeMode;   // If mode is unrecognized use COMPOSITE_CLEAR
  // the canvas accumulated thus far is the backdrop
};

// pseudo-code proposed (one less saveLayer with no need for analysis!)
          saveLayer() with setting composite mode on SkPaint
          paint Paint for src, call a)
          restore with saved composite mode
```

> the colorspace for such compositing. Tradition and existing tools specify sRGB, but there are documented advantages to linear intensity

I tend to think it makes sense to consider colorspaces in COLRv2 alongside things like CPAL variation. To get a COLRv2 we need to stick the landing on COLRv1 :)

We deliberately stuck to sRGB as that appeared to be widely and consistently supported, including W3 Compositing, and we wanted COLRv1 to be limited to such things. Looking back would we regard that as a mistake?

**Comment by dfrg** [comment]
> The way I envisioned this, there's no "underlayer". The graph is the entirety of the drawing.

This is true, but based on our reading of the spec, the intermediate nodes require the backdrop to be rendered in isolation, ignoring any previous nodes or layers that are pending composition or have already been accumulated (the “underlayer”). This seems like a useful construct to have but does carry a performance penalty and I think the question is whether it represents the common case.

**Comment by dfrg** [comment]
> Am I correct in thinking the proposed new paint looks something like this:
> 
> ```
> struct PaintComposite2
> {
>   uint8                  format; // = TBD
>   Offset24<Paint>        paint;
>   CompositeMode          compositeMode;   // If mode is unrecognized use COMPOSITE_CLEAR
>   // the canvas accumulated thus far is the backdrop
> };
> 
> // pseudo-code proposed (one less saveLayer with no need for analysis!)
>           saveLayer() with setting composite mode on SkPaint
>           paint Paint for src, call a)
>           restore with saved composite mode
> ```

This raises some additional concerns for me. For example, what is the “accumulated canvas” when this new paint is a child in a composition tree? This seems to make sense only when it is a child of `PaintColrLayers`.  Also, what does it mean if these are nested? That would imply multiple composition modes for a single source. A potential alternative might be a `PaintCompositionLayers` that associates a mode with each layer but that further complicates the structural parts of the format. 

> We deliberately stuck to sRGB as that appeared to be widely and consistently supported, including W3 Compositing, and we wanted COLRv1 to be limited to such things. Looking back would we regard that as a mistake?

Seems like a prudent decision to make at the time considering the landscape so I don’t see it as a mistake. Now might be a good time to start considering how this might be augmented in COLRv2?
